### ISSUE TITLE
Exporting spans from corpora with more than one primary text

### ISSUE THREAD

**Issue by fab-bar** [korpling/ANNIS#523]
For (historical) corpora containing multiple versions of the same text (e.g. _dipl_, _clean_, _norm_ in RIDGES), it would be nice to be able to choose one of these versions for the export of spans.

E.g., when exporting the results from the the query https://korpling.german.hu-berlin.de/annis3/?id=20d1c2e2-9400-4469-81b8-b58f15615b37 to csv, instead of an empty string for _1_span_, the user should be able to choose between _clean_, _norm_ and _dipl_ to get either „An den Leser .“ (_clean_ or _norm_) or „An den Leſer&nbsp;.“ (_dipl_). Alternatively, the export could contain all three values (_1_span_clean_, _1_span_norm_ and _1_span_dipl_).

Is this possible somehow?

Thanks,
Fabian

**Comment by thomaskrause** [comment]
This is unfortunately not possible in the moment, since the CSV exporter only exports the annotations which are part of the query and the text span. Since Ridges uses a virtual tokenization where the single primary text consists of spaces this does not work yet. I see three possible future solutions:

- We can try to extend the CSV exporter to know about segmentations. The current exporter is quite fast because it directly creates a SQL query and maps the result to CSV lines. It might get too complicated to formulate or execute such a SQL query
- Create an exporter that uses the Salt graph and export the text according to the spans. This would be slower but would not affect the existing CSV exporters
- Find a way to properly encode the segmentations as primary texts. This is something for a feature version of the ANNIS backend will support, but the "future" is still far ahead.

Currently I try to figure out if the extension of the existing CSV exporter is possible, since we have several such corpora it would be beneficial for them, too.

**Comment by fab-bar** [comment]
I have thought about it and came up with a fourth solution that works with our data (a corpus in CorA-XML format):

- Export the text from the the segmentation layer the span is attached to.

This is slightly different from what I asked about originally as the user has no choice about the text that is exported. But it assures that meaningful text is exported for spans. And it does not limit the exported spans to one specific underlying text but chooses the text according to the span. However, I am not sure if it is always the case that spans are unambiguously attached to one token layer in other data.

I have realized this solution mainly by changing the part of the SQL query returned by _selectSpanString_ in the class _annis.sqlgen.MatrixSqlGenerator_ (annis-service) to "facts.textspan AS span" and populated this column with the text that should be exported. So far, I have encountered no problems. But I have not looked very deep into the ANNIS source. And I have only tested it with our data. Do you think that there are problems with this change?

Regarding the creation of the column _textspan_:
I have browsed through the relAnnis data and it seems to me that the information to which token (or segmentation) layer a span was originally attached is not there anymore as everything is mapped directly to the layer of empty tokens. Is this correct? For my tests, I populated the column directly from the original CorA-XML data. But this is not a very nice solution.

**Comment by thomaskrause** [comment]
There is a branch (https://github.com/thomaskrause/ANNIS/tree/feature/exportdialog) which does a lot of cleanup in the exporter codes and also adds the TimelineReconstructor (https://github.com/thomaskrause/ANNIS/blob/feature/exportdialog/annis-interfaces/src/main/java/annis/TimelineReconstructor.java) which allows to get a more correct Salt graph from the database.

This could be used to write a new CSV-Exporter that is based on the reconstructed Salt Graph instead of using the matrix-query. To achieve this the CSV-Exporter must inherit from the new `SaltBasedExporter` class (https://github.com/thomaskrause/ANNIS/blob/feature/exportdialog/annis-gui/src/main/java/annis/gui/exporter/SaltBasedExporter.java) and users have to configure in their corpus properties which spans are mapped to which original tokenization (https://github.com/thomaskrause/ANNIS/blob/feature/exportdialog/annis-gui/src/main/java/annis/gui/exporter/SaltBasedExporter.java#L271).

This code is originally developed to implement a custom exporter for the Kidko project but I could do some cleanup and add this functionality to the develop branch as a base for a new CSV exporter.

**Comment by thomaskrause** [comment]
The changes which include the `SaltBasedExporter` are now in the develop branch. The exporter can be configured by adding the properties `virtual_tokenization_from_namespace` or `virtual_tokenization_mapping` to the corpus.properties file in the ExtData/ folder of the ANNIS import files.

`virtual_tokenization_from_namespace` checks if the span annotation namespace is the same as one of the tokenization names and adds a span to it if true

`virtual_tokenization_mapping` uses a comma-seperated list of "<span-anno-name>=<tokenization-name>" mappings

**Comment by thomaskrause** [comment]
Thank you for your work on this. I will check out the code later this week and try to include the changes into the next preview release.

**Comment by fab-bar** [comment]
I have added an iteration over the matches, so the exporter now produces the correct table if the matched nodes have differing numbers/types of annotations attached to them.

I also noticed that OR-queries as in #528 mess up the results - I will look at this next week.

**Comment by fab-bar** [comment]
I just realized that SaltBasedExporter changed and my CSV-Exporter does not work anymore – but making it usable again does not seem too hard.
Therefore I wanted to ask about the status of this issue. Is there any update from your side?

**Comment by thomaskrause** [comment]
I just published a preview release of the current develop branch because of another exporter we were working on. From our side the other exporter is working as expected and does not need any new functionality in the TimelineReconstructor. I had a quick look at your changes (I think the only actual conflict is some import statements) and otherwise the extension to TimelineReconstructor and the new Salt based CSV exporter are definitely  something we would like to include into  ANNIS.

**Comment by fab-bar** [comment]
I have updated the exporter so that it works again and created a pull request (#556).

As `SaltBasedExporter` now does two iterations over the matches, I use `createAdjacencyMatrix` to collect the information needed for the header (maybe this method should be given a more generic name).
For the other new methods in `SaltBasedExporter` (`getOrderedMatchNumbers` and `isAlignable`):
I did not figure out what they are supposed to do and mainly ignored them. But the exporter works with these methods simple doing nothing/ returning false.

One thing that surprised me and I want to point out to you:
the following code does not work if the `matchID` contains non-ASCII characters, `graph.getNode` will return `null`:

```
    Set<String> matchIDs = new HashSet<>(Arrays.asList(
      graph
      .getFeature(AnnisConstants.ANNIS_NS, AnnisConstants.FEAT_MATCHEDIDS)
      .getValue_STEXT().split(",")));
    for (String matchID: matchIDs)
    {
      SNode node = graph.getNode(matchID);
      …
     }

```
To make it work, `matchID` has to be decoded before calling `getNode`:

```
      matchID = URLDecoder.decode(matchID, "UTF-8");
```

I am not sure whether this is the intended behavior or not.
